Begin Chapter 06 (#21)
tidy-survey-r/tidy-survey-book: 06-descriptive-analysis-in-srvyr.Rmd

# Descriptive analyses in srvyr {#c06}

## Introduction
<!-- Definition of descriptive analytics in the context of complex survey analysis -->
Descriptive analysis describes and summarizes the basic quantitative information of survey data. Common descriptive analyses include the frequency of observations, distributions of responses, or the mean of a numeric variable. In addition to getting a sense of our data, descriptive analysis helps us identify potential errors and assess outliers. Running descriptive analysis allows us to investigate our data and gain insight into the information it contains, and it is crucial for conducting further survey analysis.
<!-- Importance of considering the design and sampling of the survey when analyzing data-->

> In complex survey analysis, we must consider the available design and sampling elements when conducting descriptive analysis. This requires reading the survey documentation (as described in Chapter 03) to apply the appropriate sampling weights, PSUs, strata, and replicate weights. Neglecting to incorporate these elements can lead to inaccurate descriptive analyses.

Reviewer: ```suggestion
In complex survey analysis, we must consider the available design and sampling elements when conducting descriptive analysis. This requires reading the survey documentation (as described in Chapter \@ref(c05)) to apply the appropriate sampling weights, PSUs, strata, and replicate weights. Neglecting to incorporate these elements can lead to inaccurate descriptive analyses.
```

The functions in the {srvyr} package allow us to run descriptive analyses while applying design and sampling elements. We can run multiple calculations in the same command thanks to the 'pipe-able' functions. The package also provides consistent return types so we can use the results with other packages like {ggplot2}.
<!-- Overview of the topics covered in the chapter -->
This chapter describes the types of descriptive analytics and the steps involved in running descriptive analysis with {srvyr}.
<!-- ## Techniques for descriptive analyses -->

```{r}
#| include: false
library(srvyr)
library(survey)
data(api)

dstrata <- apistrat %>%
  as_survey_design(strata = stype, weights = pw)
```

## Types of descriptive analyses

> Descriptive analysis can be categorized into univariate and multivariate analysis, depending on the number of variables we describe. Below, we describe the descriptive analysis for single or multiple variables and a minimal example of the {srvyr} functions associated with the measures. In the next section, we will explain each of the functions in-depth.

Reviewer 2: ```suggestion
Descriptive analysis can be categorized into univariate and multivariate analysis, depending on the number of variables we include. Below, we describe the descriptive analysis for single or multiple variables and a minimal example of the {srvyr} functions associated with the measures. In the next section, we will explain each of the functions in-depth.
```

### Univariate analysis

Univariate analysis examines a single variable at a time. A single variable has three categories of characteristics: distribution, central tendency, and dispersion. Each characteristic simplifies our dataset. Therefore, it is important to look at multiple characteristics rather than rely on a single one to describe the data.

> #### Measures of distribution

Reviewer: ```suggestion
#### Measures of distribution {-}
```

Measures of distribution describe how often an event or response occurs. These measures include proportions and quantiles.

* Examples: the proportion of students in California who did or did not receive an award based on their Academic Performance score; the proportion of U.S. citizens who voted in the last election; the distribution of household income in India divided into quantiles

The {srvyr} package includes several functions for determining measures of distribution. 

* The `survey_total()` function calculates totals
* The `survey_count()` and `survey_tally()` functions calculate weighted observations by group
* The `survey_prop()` function calculates proportions
* The `survey_quantile()` function calculates quantiles
* The `survey_ratio()` function calculates ratios

```{r}
# Calculate the survey proportion of awards

dstrata %>%
  group_by(awards) %>%
  summarise(pct = survey_prop())
```

> #### Measures of central tendency

Reviewer: ```suggestion
#### Measures of central tendency {-}
```

Measures of central tendency find the central (or average) responses. These include the mean, median, and mode.

* Examples: Average 2000 Academic Performance Index in California, the median house price in Canada

The `survey_mean()` and `survey_median()` functions from {srvyr} calculate mean and median using survey data, respectively.

```{r}
# Calculate survey mean of variable api00

dstrata %>%
  group_by(awards) %>%
  summarise(api00_mean = survey_mean(api00),
            api00_median = survey_median(api00))
```

> #### Measures of dispersion

Reviewer: ```suggestion
#### Measures of dispersion {-}
```

Measures of dispersion describe how data spreads around the central tendency. These measures include the range and standard deviation.

* Examples: The standard deviation of the 2000 Academic Performance Index in California, the range between the minimum and maximum electricity expenditure in Ohio

The `survey_var()` and `survey_sd()` functions from the {srvyr} package calculate population variance and standard deviation using survey methods, respectively.

```{r}
# Calculate the variance and standard deviation 
dstrata %>%
  summarise(api99_var = survey_var(api00),
            api99_sd = survey_sd(api00))
```

> #### A brief refresher on the `dplyr::summarize()` function

Reviewer: This section might fit better somewhere else, not sure where.

```suggestion
#### A brief refresher on the `dplyr::summarize()` function {-}
```

```{r}
mtcars %>% 
  group_by(cyl) %>% 
  summarize(mean_hp = mean(hp))
```
<!--TODO: end callout note-->

### Bivariate/multivariate analysis

Bivariate analysis concerns the relationship between two variables, while Multivariate analysis studies the relationship between multiple variables. The correlation is a number between -1 and +1 that denotes the strength of the linear relationship between two variables.

* Examples: Relationship between academic performance scores in 1999 and 2000; relationship between age and income 

The {srvyr} package does not include a function for calculating correlations for survey variables. We can use the {survey} package's `svyvar()` estimates the variance between the variables and then `cov2cor()` to get the correlation. Another option is to use the `svycor()` function from the {jtools} package.

```{r}
# Calculate correlation between API 2000 and API 1999.

var <- survey::svyvar(~api00 + api99, design = dstrata)
vmat <- as.matrix(var)
attr(vmat, "var") <- NULL # Remove extra info
attr(vmat, "statistic") <- NULL # Remove extra info
cov2cor(vmat)

> # Alternatively
> dstrata |>
>   survey::svyvar(~api00 + api99, design = _) |>
>   (\(x) as.matrix(x))() |>
>   mutate(across(everything(), var, NULL)) |>
>   (\(x) cov2cor(x))()

Reviewer: This code doesn't run for me. Maybe this instead or wait for srvyr to be updated... 

```suggestion
# Alternatively
dstrata |>
  survey::svyvar(~api00 + api99, design = _) |>
  (\(x) as.matrix(x))() |>
  (\(x) cov2cor(x))()
```

Author: This will show up when I push up my changes, but the code I have is:

```r
dstrata |>
 survey::svyvar(~api00 + api99, design = _) |>
 as.matrix(x = _) |>
 labelled::remove_attributes(x = _, "var") |>
 labelled::remove_attributes(x = _, "statistic") |>
 stats::cov2cor(V = _)
```

the output is:

```
          api00     api99
api00 1.0000000 0.9759047
api99 0.9759047 1.0000000
```

but if there's a better option, please let me know!

* Categorical/nominal data: variables with levels or descriptions that cannot be ordered, such as region of the country (North, South, East, and West)
* Ordinal data: variables that can be ordered, such as those from a Likert scale (strongly disagree, disagree, agree, and strongly agree)
* Interval/ratio: variables that are counted or measured, such as age or income
  * Within interval/ratio data are *discrete* variables, whose values are whole numbers, such as number of children, and *continuous* variables, whose values can lie anywhere on an interval, such as weight.
  
We should use different measures depending on whether our variables are interval/ratio, ordinal, or nominal.

## Descriptive analysis using the {srvyr} package

### Setup

Recall from [Chapter 05](#c05) the general process for estimation with the {srvyr} package:

1. Create a `tbl_svy` object using `srvyr::as_survey_design()` or `srvyr::as_survey_rep()`.
2. Subset the data for subpopulations using `dplyr::filter()`, if needed.
3. Specify domains of analysis using `dplyr::group_by()`, if needed.

> 4. Within `srvyr::summarize()`, specify variables to calculate ,means, totals, proportions, quantiles, and more.

Reviewer: General comment: need to introduce cascade() which is similar to summarize() but adds total rows

Need discussion on why filter needs to be done after making tbl_svy and not before.

Author: added: Filtering should be done after creating the `tbl_svy` object using `srvyr::as_survey_design()` or `srvyr::as_survey_rep()` because these functions incorporate the survey design information into the resulting object.

Please let me know if I am missing anything with that!

  wt = NULL,
  sort = FALSE,
  name = "n",
  .drop = dplyr::group_by_drop_default(x),
  vartype = c("se", "ci", "var", "cv")
  )
```

The arguments are:

* `x`: a `tbl_svy` object created by `as_survey`
* `...`: variables to group by, passed to `group_by`
* `wt`: a variable to weight on in addition to the survey weights, defaults to `NULL`
* `sort`: how to sort the variables, defaults to `FALSE`
* `name`: the name of the count variable, defaults to `n`
* `.drop`: whether to drop empty groups
* `vartype`: type(s) of variation estimate to calculate, defaults to `se` (standard error)

The steps to use `survey_count()` are:

* Specify the sample design,
* Filter subsets using `dplyr::filter()`, if needed
* Run `survey_count()`, specifying the required arguments within the function

Let's see the weighted count of responses in ANES:

```{r}
anes_des %>% #  Specify the sample design
  survey_count() # Run `survey_count()`
```

`srvyr::count()` can take one or many variables. To calculate a cross-tab of population in each age group and gender, we run the below:

```{r}
anes_des %>%
  # Specify the required arguments within the function
  survey_count(AgeGroup, Gender, name = "N") 
```

> ### Calculate totals using survey methods using `survey_total()`

Reviewer: Need an example of how this is different from `survey_count`. No good examples from ANES. For RECS, maybe something like the total amount of money spent on electricity by region?

With `srvyr::survey_total()`, we can calculate totals from complex survey data. Let's explore the syntax:

```r
survey_total(
  x,
  na.rm = FALSE,
  vartype = c("se", "ci", "var", "cv"),
  level = 0.95,
  deff = FALSE,
  df = NULL,
  ...
)
```

* `x`: a variable, expression, or empty
* `na.rm`: an indicator of whether missing values should be dropped, defaults to `FALSE`
* `vartype`: type(s) of variation estimate to calculate, defaults to `se` (standard error)
* `level`: a number or a vector indicating the confidence level, defaults to 0.95
* `deff`: a logical value stating whether the design effect should be returned, defaults to FALSE
* `df`: for `vartype = 'ci'`), a numeric value indicating degrees of freedom for the t-distribution
  * For the {srvyr} package, this defaults to `NULL` whereas the {survey} package defaults to `Inf`

The steps to use `survey_total()` are:

* Specify the sample design,
* Specify the cross tab in `group_by()`,
* Within `summarize`, run `survey_total()`, specifying the required arguments within the function

To calculate a population count estimate with `survey_total()`, we can run the below:

```{r}

> anes_des %>%
>   summarize(survey_total(), .groups = "drop")

Reviewer: Assign the result to something.

```suggestion
anes_des %>%
  summarize(N=survey_total(), .groups = "drop")

```

```

The `.groups =` argument controls the grouping structure of the output. When the output no longer have grouping variables because they are dropped, it becomes ungrouped. <!-- TODO: Check this after dplyr 1.1.0 is released -->

Notice that `anes_des %>% summarize(survey_total(), .groups = "drop")` is equivalent to the `survey_count()` call:

```{r}
anes_des %>%
  survey_count() 
```

The `survey_total()` function is called within `summarize`, where as `survey_count()`, like `dplyr::count()`, is not.

### Calculate mean/proportion using survey methods with `survey_mean()` and `survey_prop()`

> The `srvyr::survey_mean()` and `survey_prop()` functions calculate the means and proportions from complex survey data. Like `survey_total()`, they are called within `summarize()`. Let's explore the syntax:

Reviewer: Bring in an example with survey_mean. Need something like RECS with continuous outcomes.
